irohad/torii/processor/impl/transaction_processor_impl.cpp

```diff
@@ -49,49 +51,79 @@ namespace iroha {
         }
       });
 
-      // move commited txs from proposal to candidate map
+      // process stateful validation results
+      pcs_->on_verified_proposal().subscribe(
+          [this](std::shared_ptr<validation::VerifiedProposalAndErrors>
+                     proposal_and_errors) {
+            // notify about failed txs
+            auto errors = proposal_and_errors->second;
+            std::lock_guard<std::mutex> lock(notifier_mutex_);
+            for (const auto &tx_error : errors) {
+              std::string error_msg;
+              if (not tx_error.first.tx_passed_initial_validation) {
```

Comment:
Maybe move this whole section about construction of error message to a separate function